tsdate/evaluation.py

```diff
@@ -1,254 +1,354 @@
+    matched_time[matched_span == 0] = np.nan
+
+    return matched_time, matched_span, best_match
+
+
+def node_spans(ts):
+    """
+    Returns the array of "node spans", i.e., the `j`th entry gives
+    the total span over which node `j` is in the tree (i.e., does
+    not have 'missing data' there).
+    """
+    child_spans = np.bincount(
+        ts.edges_child,
+        weights=ts.edges_right - ts.edges_left,
+        minlength=ts.num_nodes,
+    )
+    for t in ts.trees():
+        span = t.span
+        for r in t.roots:
+            # do this check to exempt 'missing data'
+            if t.num_children(r) > 0:
+                child_spans[r] += span
+    return child_spans
+
+
+def tree_discrepancy(ts, other):
+    """
+    For two tree sequences `ts` and `other`,
+    this method returns two values, as a tuple:
+    1. the fraction of the total span of `ts` over which each nodes' descendant
+    sample set does not match its' best match's descendant sample set.
+    2. The root mean squared difference
+    between the times of the nodes in `ts`
+    and times of their best matching nodes in `other`,
+    with the average weighted by the nodes' spans in `ts`.
+
+    This is done as follows:
+
+    For each node in `ts` the best matching node(s) from `other`
+    has the longest matching span using `shared_node_spans`.
+    If there are multiple matches with the same longest shared span
+    for a single node, the best match is the match that is closest in time.
+    The discrepancy is:
+    ..math::
+
+    d(ts, other) = 1 -
+    \\left(sum_{x\\in \\operatorname{ts}}
+    \\min_{y\\in \\operatorname{other}}
+    |t_x - t_y| \\max{y \\in \\operatorname{other}}
+    \frac{1}{T}* \\operatorname{shared_span}(x,y)\right),
+
+    where :math: `T` is the sum of spans of all nodes in `ts`.
+
+    Returns two values:
+    `discrepancy` (float) the value computed above.
+    `root-mean-squared discrepancy` (float)
+    """
+
+    shared_spans = shared_node_spans(ts, other)
+    # Find all potential matches for a node based on max shared span length
+    max_span = shared_spans.max(axis=1).toarray().flatten()
+    col_ind = shared_spans.indices
+    row_ind = np.repeat(
+        np.arange(shared_spans.shape[0]), repeats=np.diff(shared_spans.indptr)
+    )
+
+    match = shared_spans.data == max_span[row_ind]
+    # Construct a matrix of potiential matches and
+    match_matrix = scipy.sparse.coo_matrix(
+        (shared_spans.data[match], (row_ind[match], col_ind[match])),
+        shape=(ts.num_nodes, other.num_nodes),
+    )
+    # scale with difference in node times
+    # determine best matches with the best_match_matrix
+    ts_times = ts.nodes_time[row_ind[match]]
+    other_times = other.nodes_time[col_ind[match]]
+    time_difference = np.absolute(np.asarray(ts_times - other_times))
+    best_match_matrix = scipy.sparse.coo_matrix(
+        (
+            1 / (1 + (match_matrix.data * time_difference)),

```

Comment:
Same as earlier comment -- couldn't we end up with best matches that don't satisfy the "max shared span" criterion, but are very similar in terms of age?

And, the choice of time scaling would impact the chance of this happening, right? In that, if the time scaling is such that the absolute difference between times is large relative to typical node spans, then this one-step criterion would be dominated by the difference in ages?

Reply:
Nevermind, I see how this works-- very nice.